cmds/switch_root/switch_root.go

```diff
@@ -23,140 +24,201 @@ func usage() string {
 	return "switch_root [-h] [-V]\nswitch_root newroot init"
 }
 
-// littleDoctor recursively deletes everything at "path" that
-// is in the same file system as "fs".
-func littleDoctor(path string, fsType int64) error {
-	var pathFS syscall.Statfs_t
+// get device for file descriptor
+func getDev(fd int) (dev uint64, err error) {
+	var stat syscall.Stat_t
 
-	if err := syscall.Statfs(path, &pathFS); err != nil {
-		return err
+	if err := syscall.Fstat(fd, &stat); err != nil {
+		return 0, err
 	}
 
-	if int64(pathFS.Type) != fsType {
-		return nil
-	}
+	return stat.Dev, nil
+}
 
-	file, err := os.Open(path)
-	if err != nil {
-		return fmt.Errorf("Could not open %s: %v", path, err)
+// check for the symlink bit
+func isSymlink(mode uint32) bool {
+	// numbers come from `man fstatat`
+	return (mode & 0170000) == 0120000
+}
+
+// recursively deletes a directory and everything in it
+// works with a file descriptor, can delete files not referenceable
+// any more (e.g. after chroot)
+// does not descent into mounts
+func recursiveDelete(fd int) error {
+	var rb syscall.Stat_t
+
+	if err := syscall.Fstat(fd, &rb); err != nil {
+		return err
 	}
 
-	if fileStat, err := file.Stat(); fileStat.IsDir() {
-		if err != nil {
-			return err
-		}
+	// may need to call syscall.ReadDirent multiple times
+	for {
 
-		names, err := file.Readdirnames(-1)
-		defer file.Close()
+		// get the subfiles in buff
+		buff := make([]byte, 4096)
+		var nbuff int
+		nbuff, err := syscall.ReadDirent(fd, buff)
 		if err != nil {
 			return err
 		}
-
-		for _, fileName := range names {
-
-			if fileName == "." || fileName == ".." {
-				continue
-			}
-
-			if err := littleDoctor(filepath.Join(path, fileName), fsType); err != nil {
-				return err
-			}
-			if err := os.Remove(path); err != nil {
-				return err
-			}
+		if nbuff <= 0 {
+			break
 		}
 
-	} else {
-		if err := os.Remove(path); err != nil {
-			return err
+		_, _, names := syscall.ParseDirent(buff, nbuff, []string{})
+
+		for _, name := range names {
+			// try to open it with O_DIRECTORY to know if it's a directory
+			// NOTE O_DIRECTORY is linux specific
+			namefd, err := syscall.Openat(fd, name, syscall.O_DIRECTORY, syscall.O_RDWR)
+			if err != nil {
+				// just delete files
+				if err := unlinkat(fd, name, 0); err != nil {
+					return err
+				}
+			} else {
+				defer syscall.Close(namefd)
```

Comment:
This is a bit dangerous. You're going to stack up defers of *every* file you open until the function returns, which means that all files deleted by this function won't actually get deleted until you return. That'll be a lot of defers. Besides that, you could easily run out of file descriptors (`RLIMIT_NOFILE`) with this.

I'd recommend putting the logic inside the for-loop here in its own function -- that way the fd is closed on return.

Reply:
Aaaah, I misread the documentation and somehow assumed defer would be run before the next loop iteration (when the C-style "scope" of `namefd` ends). Will move to own function in this case. :thumbsup: